contracts/ConverterRamp.sol

```diff
@@ -0,0 +1,383 @@
+        }
+    }
+
+    function convertSafe(
+        TokenConverter converter,
+        Token fromToken,
+        Token toToken,
+        uint256 amount
+    ) internal returns (uint256 bought) {
+        if (address(fromToken) != ETH_ADDRESS) require(fromToken.approve(address(converter), amount), 'Error approving token transfer');
+        uint256 prevBalance = address(toToken) != ETH_ADDRESS ? toToken.balanceOf(address(this)) : address(this).balance;
+        uint256 sendEth = address(fromToken) == ETH_ADDRESS ? amount : 0;
+        uint256 boughtAmount = converter.convert.value(sendEth)(fromToken, toToken, amount, 1);
+        require(
+            boughtAmount == (address(toToken) != ETH_ADDRESS ? toToken.balanceOf(address(this)) : address(this).balance) - prevBalance,
+            'Bought amound does does not match'
+        );
+        if (address(fromToken) != ETH_ADDRESS) require(fromToken.approve(address(converter), 0), 'Error removing token approve');
+        return boughtAmount;
+    }
+
+    function executeOptimalPay(
+        bytes32[5] memory params,
+        bytes memory oracleData,
+        uint256 rcnToPay
+    ) internal returns (bool) {
+        LoanManager loanManager = LoanManager(address(uint256(params[I_LOAN_MANAGER])));
+        bytes32 id = params[I_REQUEST_ID];
+        RateOracle oracle = RateOracle(loanManager.getOracle(uint256(id)));
+
+        uint256 toPay;
+
+        if (address(oracle) == address(0)) {
+            toPay = rcnToPay;
+        } else {
+            (uint256 _tokens, uint256 _equivalent) = oracle.readSample(oracleData);
+            require(_equivalent != 0, 'Oracle provided invalid rate');
+            toPay = rcnToPay.mul(_equivalent) / _tokens;
+        }
+
+        Token rcn = loanManager.token();
+        require(rcn.approve(address(loanManager), rcnToPay), 'Error on payment approve');
+
+        DebtEngine debtEngine = DebtEngine(address(uint256(params[I_DEBT_ENGINE])));
+        debtEngine.pay(id, toPay, address(uint256(params[I_PAY_FROM])), oracleData); //VERIFICAR ALGO ?
+        require(rcn.approve(address(loanManager), 0), 'Error removing the payment approve');
+
+        return true;
+    }
+
+    function executeLend(
+        bytes32[4] memory params,
+        bytes memory oracleData,
+        bytes memory cosignerData
+    ) internal returns (bool) {
+        LoanManager loanManager = LoanManager(address(uint256(params[I_LOAN_MANAGER])));
+        bytes32 id = params[I_REQUEST_ID];
+        return loanManager.lend(id, oracleData, address(uint256(params[I_LEND_COSIGNER])), 0, cosignerData);
+    }
+
+    function executeTransfer(
+        bytes32[4] memory params,
+        address to
+    ) internal returns (bool) {
+        DebtEngine debtEngine = DebtEngine(address(uint256(params[I_DEBT_ENGINE])));
+        // TODO: ANALIZAR ESTO! debtEngine.safeTransferFrom(, to, uint256(params[I_REQUEST_ID]));
```

Comment:
tenemos que ver esto

Reply:
To have better compatibility we can use `transferFrom` instead of `safeTransferFrom`, I don't fully understand the question, the end result should be something like:

```solidity
debtEngine.safeTransferFrom(address(this), to, uint256(params[I_REQUEST_ID]));
```